# Lập trình cho Khoa học dữ liệu - Đồ án cuối kì

18120198 - Nguyễn Huỳnh Đại Lợi

18120212 - Huỳnh Long Nam

---

## Import

In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd 

## Thu thập dữ liệu

Dữ liệu được trích từ bộ dữ liệu [Mekong Infrastructure Tracker](https://www.stimson.org/project/mekong-infrastructure/). Bộ dữ liệu cho ta một cái nhìn về cơ sở hạ tầng cho năng lượng, giao thông, nguồn nước ở đồng bằng sông Mekong.

Trong bài này, chúng em sử dụng dữ liệu về Việt Nam thôi nên sẽ sử dụng filter có sẵn của nguồn để điều chỉnh chỉ lấy dữ liệu các cơ sở của Việt Nam. Dữ liệu được tải qua công cụ tracker của nguồn [ở đây](https://www.stimson.org/2020/mekong-infrastructure-tracker-tool/).

File `mekong-power-generation.xlsx` được đính kèm là file đã được tải về(lần cuối tải về là ngày 7/5/2021) và là dữ liệu sẽ được sử dụng.

---

## Khám phá dữ liệu

### Đọc dữ liệu từ file

In [2]:
df=pd.read_excel("mekong-power-generation.xlsx")
df.head()

Project Name Subtype Current Status  \
0                                             A Luoi   Hydro    Operational   
1                                            A Vuong   Hydro    Operational   
2                                              Amata     Oil    Operational   
3  An Khanh - Bac Giang (Nhà máy Nhiệt điện An Kh...    Coal        Planned   
4         An Khanh I (Nhà máy nhiệt điện An Khánh I)    Coal    Operational   

   Capacity (MW)  Year of Completion  undefined  \
0          170.0              2012.0        NaN   
1          210.0              2008.0        NaN   
2           13.0                 NaN        NaN   
3          650.0              2023.0        NaN   
4          120.0              2015.0        NaN   

  Country list of Sponsor/Developer               Sponsor/Developer Company  \
0                           Vietnam           Central Region Hydropower JSC   
1                           Vietnam                 A Vuong Hydro Power JSC   
2                               NaN                                     NaN   
3                           Vietnam  An Khanh - Bac Giang Thermal Power JSC   
4                           Vietnam                An Khanh Electricity JSC   

  Country list of Lender/Financier  \
0                              NaN   
1                              NaN   
2                              NaN   
3                            China   
4                   China; Vietnam   

                            Lender/Financier Company  ...  \
0                                                NaN  ...   
1                                                NaN  ...   
2                                                NaN  ...   
3  China Development Bank (CDB); Industrial and C...  ...   
4  China Communications Bank; Bank of China, Chin...  ...   

                           Tributary  Latitude Longitude  \
0                             Sekong  16.19762  107.1619   
1                         Song Giang  15.79980  107.6171   
2                           Dong Nai  10.94710  106.9000   
3  Song Cau-Song Thuong-Song Luc Nam  21.23212  106.3596   
4  Song Cau-Song Thuong-Song Luc Nam  21.59691  105.7666   

                                           Proximity Avg. Annual Output (MWh)  \
0                                                                               
1                                                                               
2                                                                               
3  Actual location unknown, currently sited at lo...                  5000000   
4                                                                      800000   

   undefined.2                Data Source  \
0          NaN        Vietnam Electricity   
1          NaN                  Wikipedia   
2          NaN  World Resources Institute   
3          NaN            Sponsor Website   
4          NaN                    AidData   

                       Announcement/More Information  \
0  https://en.evn.com.vn/userfile/User/huongbtt/f...   
1  https://www.licogi166.vn/en/cong-trinh-thuy-di...   
2  http://www.evn.com.vn/userfile/files/2017/3/An...   
3  http://danviet.vn/doanh-nghiep/nhiet-dien-an-k...   
4  https://vietnamnews.vn/economy/219017/an-khanh...   

                                                Link Latest Update  
0  https://wle-mekong.cgiar.org/construction-of-n...    03/31/2021  
1                                                       03/31/2021  
2                                                                   
3                                                                   
4  http://nhietdienankhanh.com.vn/gioi-thieu/gioi...    03/29/2021  

[5 rows x 26 columns]

### Kích thước dữ liệu

In [3]:
shape = df.shape
shape

(563, 26)

### Tìm hiểu về các thuộc tính của dữ liệu sử dụng

Dữ liệu gồm:
- 563 dòng tương ứng với 551 dự án năng lượng tại Việt Nam
- 26 cột, gồm các cột sau:

In [4]:
df.columns.to_list()

['Project Name',
 'Subtype',
 'Current Status',
 'Capacity (MW)',
 'Year of Completion',
 'undefined',
 'Country list of Sponsor/Developer',
 'Sponsor/Developer Company',
 'Country list of Lender/Financier',
 'Lender/Financier Company',
 'Country list of Construction/EPC',
 'Construction Company/EPC Participant',
 'Country',
 'Province/State',
 'District',
 'undefined.1',
 'Tributary',
 'Latitude',
 'Longitude',
 'Proximity',
 'Avg. Annual Output (MWh)',
 'undefined.2',
 'Data Source',
 'Announcement/More Information',
 'Link',
 'Latest Update']

Xét bộ dữ liệu tải về, các cột undefined/undefined.1/undefined.2 là các cột vô nghĩa(có thể là lỗi của nguồn dữ liệu), theo sự tìm hiểu của nhóm thì các cột có ý nghĩa mang nội dung sau:
- Project Name: tên dự án
- Subtype: loại năng lượng (than, thủy điện, mặt trời, sinh học,...)
- Current Status: tình trạng hiện tại (lên kế hoạch, đang thi công, đã bị hủy,...)
- Capacity (MW): công suất 
- Year of Completion: năm hoàn thành
- Country list of Sponsor/Developer: các quốc gia tài trợ, phát triển
- Sponsor/Developer Company: các công ty tài trợ, phát triển
- Country list of Lender/Financier: các quốc gia cho vay 
- Lender/Financier Company: các công ty cho vay 
- Country list of Construction/EPC*: các quốc gia tham gia vào xây dựng, tổng thầu xây dựng
- Construction Company/EPC Participant: công ty,tổng thầu xây dựng
- Country: dự án đặt tại quốc gia nào
- Province/State: dự án đặt tại tỉnh/bang nào
- District: dự án đặt tại huyện nào 
- Tributary: phụ lưu của dự án
- Latitude: vĩ độ
- Longitude: kinh độ
- Avg. Annual Output (MWh): sản lượng hàng năm
- Data Source: nguồn dữ liệu
- Announcement/More Information: Link thông tin dự án
- Link: Link thông tin dự án
- Latest Update: Lần cuối cập nhật


 `*EPC (viết tắt của từ Engineering, Procurement and Construction) có nghĩa là xây dựng một hợp đồng xây dựng mà nhà thầu thực hiện toàn bộ các công việc kỹ thuật, cung ứng vật tư, thiết bị đến việc thi công xây dựng công trình, hạng mục và chạy thử nghiệm bàn giao cho chủ đầu tư`

### Kiểm tra dữ liệu có bị lặp không

In [5]:
df[df.duplicated()]

Empty DataFrame
Columns: [Project Name, Subtype, Current Status, Capacity (MW), Year of Completion, undefined, Country list of Sponsor/Developer, Sponsor/Developer Company, Country list of Lender/Financier, Lender/Financier Company, Country list of Construction/EPC, Construction Company/EPC Participant, Country, Province/State, District, undefined.1, Tributary, Latitude, Longitude, Proximity, Avg. Annual Output (MWh), undefined.2, Data Source, Announcement/More Information, Link, Latest Update]
Index: []

[0 rows x 26 columns]

Dữ liệu có vẻ không có dòng nào bị trùng với nhau

### Kiểm tra kiểu dữ liệu mỗi thuộc tính và xử lí các thuộc tính có vấn đề về kiểu dữ liệu

In [6]:
dtypes = df.dtypes
dtypes

Project Name                             object
Subtype                                  object
Current Status                           object
Capacity (MW)                           float64
Year of Completion                      float64
undefined                               float64
Country list of Sponsor/Developer        object
Sponsor/Developer Company                object
Country list of Lender/Financier         object
Lender/Financier Company                 object
Country list of Construction/EPC         object
Construction Company/EPC Participant     object
Country                                  object
Province/State                           object
District                                 object
undefined.1                             float64
Tributary                                object
Latitude                                float64
Longitude                               float64
Proximity                                object
Avg. Annual Output (MWh)                

Thuộc tính `Avg. Annual Output (MWh)` thuộc kiểu dữ liệu object trong khi phải là kiểu dữ liệu số

In [7]:
df['Avg. Annual Output (MWh)'].unique()

array([' ', '5000000', '800000', '694000', nan, '639000', '80000',
       '70000', '100000', '60000', '250000', '545000', '600000', '84000',
       '77000', '64800', '44000', '74000', '12500', '10000', '73904',
       '378000', '120000', '94710', '25000', '32000', '56300', '76000',
       '13000', '93200', '217642', '220500', '752510', '69000', '53000',
       '65000', '11400', '69100', '15780', '115000', '636800', '143000',
       '40000', '105000', '68000', '85800', '37000', '75000', '110000',
       '338000', '688000', '45000', '66003', '161000', '720000', '72820',
       '4000000', '1350000', '1930000', '3900000', '140000', '82000',
       '63650', '31860', '30000', '73000', '55000', '50800', '8160000',
       '280000', '334148', '132000', '1904000', '123329', '81135',
       '965000', '17500', '328000', '21000', '105870', '63500', '4690000',
       '22036', '67630', '28000', '54000', '79000', '46300', '19000',
       '18000', '12000', '59000', '56000', '127500', '48850', '146000',

Có vẻ trong bộ dữ liệu có pha lẫn giá trị khoảng trắng '$~~$' thay vì rỗng. Ta kiểm tra còn thuộc tính nào khác chứa chuỗi '$~~$' nữa không

In [8]:
space_count=df.isin([' ']).sum(axis=0)
space_count

Project Name                              0
Subtype                                   0
Current Status                            0
Capacity (MW)                             0
Year of Completion                        0
undefined                                 0
Country list of Sponsor/Developer         0
Sponsor/Developer Company                 0
Country list of Lender/Financier          0
Lender/Financier Company                  0
Country list of Construction/EPC          0
Construction Company/EPC Participant    387
Country                                   0
Province/State                            0
District                                  0
undefined.1                               0
Tributary                                 0
Latitude                                  0
Longitude                                 0
Proximity                               366
Avg. Annual Output (MWh)                311
undefined.2                               0
Data Source                     

---

## Tiền xử lý

Ta thực hiện tiền xử lý để thay các giá trị chuỗi khoảng trắng ' ' thành giá trị `nan` và chuyển cột `Avg. Annual Output (MWh)` thành dạng `float`

In [9]:
columns_that_contains_spaces=space_count[space_count>0].index
df.loc[:,columns_that_contains_spaces]=df.loc[:,columns_that_contains_spaces].replace({' ':np.nan})
df['Avg. Annual Output (MWh)']=df['Avg. Annual Output (MWh)'].astype(float)

Sau khi thay các khoảng trống ấy bằng giá trị `nan` thì đã không còn vấn đề khoảng trắng cũng như về kiểu dữ liệu

In [10]:
df.isin([' ']).sum(axis=0)

Project Name                            0
Subtype                                 0
Current Status                          0
Capacity (MW)                           0
Year of Completion                      0
undefined                               0
Country list of Sponsor/Developer       0
Sponsor/Developer Company               0
Country list of Lender/Financier        0
Lender/Financier Company                0
Country list of Construction/EPC        0
Construction Company/EPC Participant    0
Country                                 0
Province/State                          0
District                                0
undefined.1                             0
Tributary                               0
Latitude                                0
Longitude                               0
Proximity                               0
Avg. Annual Output (MWh)                0
undefined.2                             0
Data Source                             0
Announcement/More Information     

In [11]:
df.dtypes

Project Name                             object
Subtype                                  object
Current Status                           object
Capacity (MW)                           float64
Year of Completion                      float64
undefined                               float64
Country list of Sponsor/Developer        object
Sponsor/Developer Company                object
Country list of Lender/Financier         object
Lender/Financier Company                 object
Country list of Construction/EPC         object
Construction Company/EPC Participant     object
Country                                  object
Province/State                           object
District                                 object
undefined.1                             float64
Tributary                                object
Latitude                                float64
Longitude                               float64
Proximity                                object
Avg. Annual Output (MWh)                

---

## Quay lại bước khám phá dữ liệu 

### Khám phá các cột giá trị số

In [12]:
num_col_info_df = pd.DataFrame(columns=df.columns[df.dtypes == np.number], index=["missing_ratio", "min", "lower_quartile", "median", "upper_quartile", "max"])
num_col_info_df.drop(columns=['undefined','undefined.1','undefined.2'],inplace=True)

In [28]:
df_num=df[num_col_info_df.columns]
num_col_info_df.loc['missing_ratio'] = (df_num.isna().sum()/len(df_num.index)*100).round(1)
num_col_info_df.loc['min'] = df_num.min().round(1)
num_col_info_df.loc['max'] = df_num.max().round(1)
num_col_info_df.loc['lower_quartile'] = df_num.quantile(0.25).round(1)
num_col_info_df.loc['upper_quartile'] = df_num.quantile(0.75).round(1)
num_col_info_df.loc['median'] = df_num.median().round(1)
num_col_info_df

Capacity (MW) Year of Completion Latitude Longitude  \
missing_ratio            2.8               11.0      0.0       0.0   
min                      0.2             1945.0      8.5     102.5   
lower_quartile          24.0             2012.0     11.5     105.7   
median                  50.0             2018.0     14.0     107.0   
upper_quartile         269.5             2020.0     20.5     108.3   
max                   9583.0             2031.0     23.2     109.3   

               Avg. Annual Output (MWh)  
missing_ratio                      61.1  
min                              7000.0  
lower_quartile                  60000.0  
median                          80000.0  
upper_quartile                 213321.0  
max                          12000000.0

Các giá trị thiếu tập trung chủ yếu ở cột `Avg. Annual Output (MWh)` nên ta có thể drop cột đó

In [14]:
cols_to_drop=['Avg. Annual Output (MWh)','undefined', 'undefined.1', 'undefined.2']

### Khám phá các cột giá trị không phải số

In [16]:
notnum_col_info_df = pd.DataFrame(columns=df.columns[df.dtypes != np.number], index=["missing_ratio", "num_values", "value_ratios"])
df_notnum=df[notnum_col_info_df.columns]

In [25]:
multichoice_cols=['Country list of Sponsor/Developer', 'Sponsor/Developer Company' ,'Country list of Lender/Financier','Country list of Construction/EPC','Construction Company/EPC Participant','Data Source']

In [28]:
notnum_col_info_df.loc['missing_ratio'] = (df_notnum.isna().sum()/len(df_notnum.index)*100).round(1)
for c in notnum_col_info_df.columns:
    di=dict()
    col_total_value=len(df_notnum[c].index)-df_notnum[c].isna().sum()
    if c in multichoice_cols:
        li=[]
        for i in df_notnum[c]:
            if type(i)==str:
                li+=i.split(";")
        temp_set=set(li)
        notnum_col_info_df[c]['num_values']=len(temp_set)
        for i in temp_set:
            value_appearance=li.count(i)*100
            di[i]=(value_appearance/len(li))
        for dict_value in di:
            for k, v in di.items():
                di[k] = round(v, 1)
        notnum_col_info_df[c]['value_ratios']=dict(sorted(di.items(), key=lambda x: x[1],reverse=True))
    else:
        notnum_col_info_df[c]['num_values']=df_notnum[c].nunique()
        value_set=df_notnum[c].unique()
        for i in value_set:
            if type(i)==str:
                di[i]=((list(df_notnum[c]).count(i)/col_total_value)*100).round(1)
        notnum_col_info_df[c]['value_ratios']=dict(sorted(di.items(), key=lambda x: x[1],reverse=True))
        

notnum_col_info_df

Project Name  \
missing_ratio                                                0.0   
num_values                                                   561   
value_ratios   {'Cam An Bắc (Nhà máy điện mặt trời xã Cam An ...   

                                                         Subtype  \
missing_ratio                                                0.0   
num_values                                                     8   
value_ratios   {'Hydro': 43.7, 'Solar': 22.4, 'Coal': 14.4, '...   

                                                  Current Status  \
missing_ratio                                                0.0   
num_values                                                     6   
value_ratios   {'Operational': 65.5, 'Planned': 19.9, 'Under ...   

                               Country list of Sponsor/Developer  \
missing_ratio                                                7.1   
num_values                                                    38   
value_ratios   {'Vietnam': 77.6, ' Vietnam': 3.2, 'Thailand':...   

                                       Sponsor/Developer Company  \
missing_ratio                                                6.2   
num_values                                                   388   
value_ratios   {'EVN': 7.3, 'EVN (Vietnam)': 2.9, 'Vietnam El...   

                                Country list of Lender/Financier  \
missing_ratio                                               82.2   
num_values                                                    31   
value_ratios   {'Vietnam': 20.4, 'Japan': 16.1, 'China': 10.9...   

                                        Lender/Financier Company  \
missing_ratio                                               82.4   
num_values                                                    70   
value_ratios   {'JICA': 9.1, 'USTDA': 6.1, 'World Bank Group'...   

                                Country list of Construction/EPC  \
missing_ratio                                               73.9   
num_values                                                    26   
value_ratios   {'Vietnam': 35.8, 'China': 21.2, 'Japan': 10.6...   

                            Construction Company/EPC Participant  \
missing_ratio                                               73.0   
num_values                                                   133   
value_ratios   {'First Solar (Equipment)': 3.8, 'Cty CP năng ...   

                          Country  \
missing_ratio                 0.0   
num_values                      1   
value_ratios   {'Vietnam': 100.0}   

                                                  Province/State  \
missing_ratio                                                0.0   
num_values                                                    60   
value_ratios   {'Ninh Thuận': 8.3, 'Bình Thuận': 5.5, 'Sơn La...   

                                                        District  \
missing_ratio                                                0.0   
num_values                                                   228   
value_ratios   {'Thuận Nam': 2.7, 'Mường La': 2.7, 'Hướng Hóa...   

                                                       Tributary  \
missing_ratio                                                1.1   
num_values                                                    60   
value_ratios   {'Black-Li Xian': 10.2, 'Vietnam Mekong Delta'...   

                                                       Proximity  \
missing_ratio                                               69.4   
num_values                                                     2   
value_ratios   {'Actual location unknown, currently sited at ...   

                                                     Data Source  \
missing_ratio                                                1.2   
num_values                                                    37   
value_ratios   {'Open Development Vietnam': 21.5, 'Sponsor We...   

                                   Announcement/More Information  \
missing_ratio